### DISCUSSION TITLE
Get the  "ERROR Cannot find page" on build

### DISCUSSION THREAD

**Discussion by vinser** [gethinode/hinode#934] [Help] [answered]
Hi

When I run `hugo` command to build my site I get errors like these
```log
ERROR Cannot find page: 'https://vinser.github.io/burnfix' at "/home/vinser/Dev/Sites/hugo-hinode/content/en/apps/burnfix.md:23:61"
ERROR Cannot find page: 'https://vinser.github.io/flibgolite' at "/home/vinser/Dev/Sites/hugo-hinode/content/en/apps/flibgolite.md:27:61"
ERROR Cannot find page: 'https://vinser.github.io/pixyne' at "/home/vinser/Dev/Sites/hugo-hinode/content/en/apps/pixyne.md:26:61"
```
And when I run `hugo server`  command it builds the site without any flaws and the site runs well 

In my `hugo.toml` I've set 
```toml
baseURL = 'https://vinser.github.io/'
```
and links are as above mentioned in errors
```toml
[links]
    burnfix_docs = "https://vinser.github.io/burnfix"
    flibgolite_docs = "https://vinser.github.io/flibgolite"
    pixyne_docs = "https://vinser.github.io/pixyne"
```

If I change link ref from `https://vinser.github.io/burnfix'` for ex to `https://github.com/vinser/burnfix` build ends well

It seems that hugo when doing product build consider such links are on the same host or so and is trying to find local ref?

Please does anybody knows a workaround of such a case

**Comment by markdumay** [comment]
Can you share a bit more context and/or explain the steps how to reproduce this? Is this an error with a shortcode (e.g. [link](https://gethinode.com/docs/components/link/), [navbar](https://gethinode.com/docs/components/navbar/))?

**Comment by vinser** [comment]
The code is here https://github.com/vinser/hugo-hinode. It has home and 3 app pages - just light enough to test the case
To reproduce:
1. git clone the repo
2. hugo - to see errors
3. hugo server - to see site

**Comment by vinser** [comment]
That's how it looks:
```
~/Dev/Sites/hugo-hinode$ hugo
Start building sites … 
hugo v0.126.1-3d40aba512931031921463dafc172c0d124437b8+extended linux/arm64 BuildDate=2024-05-15T10:42:34Z VendorInfo=gohugoio

ERROR Cannot find page: 'https://vinser.github.io/flibgolite' at "/home/vinser/Dev/Sites/hugo-hinode/content/en/apps/flibgolite.md:27:61"
ERROR Cannot find page: 'https://vinser.github.io/pixyne' at "/home/vinser/Dev/Sites/hugo-hinode/content/en/apps/pixyne.md:26:61"
ERROR Cannot find page: 'https://vinser.github.io/burnfix' at "/home/vinser/Dev/Sites/hugo-hinode/content/en/apps/burnfix.md:23:61"
ERROR Cannot find page: 'https://vinser.github.io/burnfix' at "/home/vinser/Dev/Sites/hugo-hinode/content/en/apps/burnfix.md:21:22"
ERROR Cannot find page: 'https://vinser.github.io/pixyne' at "/home/vinser/Dev/Sites/hugo-hinode/content/en/apps/pixyne.md:23:276"
ERROR Cannot find page: 'https://vinser.github.io/flibgolite' at "/home/vinser/Dev/Sites/hugo-hinode/content/en/apps/flibgolite.md:24:53"
Total in 1763 ms
Error: error building site: logged 6 error(s)

~/Dev/Sites/hugo-hinode$ hugo server
Watching for changes in /home/vinser/Dev/Sites/hugo-hinode/{archetypes,assets,content,data,i18n,layouts,static}
Watching for config changes in /home/vinser/Dev/Sites/hugo-hinode/hugo.toml, /home/vinser/Dev/Sites/hugo-hinode/config/_default, /home/vinser/Dev/Sites/hugo-hinode/go.mod
Start building sites … 
hugo v0.126.1-3d40aba512931031921463dafc172c0d124437b8+extended linux/arm64 BuildDate=2024-05-15T10:42:34Z VendorInfo=gohugoio


                   | EN   
-------------------+------
  Pages            |  15  
  Paginator pages  |   0  
  Non-page files   |   0  
  Static files     | 118  
  Processed images |  67  
  Aliases          |   2  
  Cleaned          |   0  

Built in 1399 ms
Environment: "development"
Serving pages from disk
Running in Fast Render Mode. For full rebuilds on change: hugo server --disableFastRender
Web Server is available at http://localhost:1313/ (bind address 127.0.0.1) 
Press Ctrl+C to stop
```

**Comment by vinser** [comment]
Pure Markdown link 
```
[product website ](https://vinser.github.io/burnfix/)
```
instead of 
```
{{< link href="https://vinser.github.io/burnfix/" >}}product website{{< /link >}}
``` 
does not cause this error

**Comment by vinser** [comment]
Based on the above, I can assume that if Hinode link contains external ref/url that extends the baseURL, then Hinode assigns it the attribute of an internal/local link and, when building, hugo cannot find it.

Should I open the issue?

**Comment by markdumay** [comment]
I'll look into it!

As a first observation, I'd recommend to use relative links instead of absolute links. For example:

```
Absolute:
{{< link href="https://vinser.github.io/burnfix/" >}}product website{{< /link >}} 

Relative link:
{{< link href="burnfix" >}}product website{{< /link >}} 
```

**Reply by markdumay** [reply]
I published [Hinode v0.23.20 ](https://github.com/gethinode/hinode/releases/tag/v0.23.20)that improves the handling of links. The build no longer chokes on links that include the site's [base URL](https://gohugo.io/methods/site/baseurl/).

You're still encouraged to use relative links though. Using absolute links makes your site less portable. Additionally, the base URL is not consistent across build environments. For example, in Hugo's server mode, the base URL is replaced with a local address (e.g. `http://localhost:1313/`). As a result, Hinode treats `https://vinser.github.io/burnfix/` as an external address (adding an icon and opening in a new tab by default), although it should really be a local link. The same applies when using build previews, as typically these previews use an alternative base URL.

**Reply by vinser** [reply]
Many thanks for your efforts to make the Hignode theme the best but to my great regret, the v0.23.20 fix does not help.
Errors are the same
```
~/Dev/Sites/hugo-hinode/site$ hugo mod get -u ./...
Update module in /home/vinser/Dev/Sites/hugo-hinode/site
go: downloading github.com/gethinode/hinode v0.23.20
go: upgraded github.com/gethinode/hinode v0.23.19 => v0.23.20

~/Dev/Sites/hugo-hinode/site$ hugo
Start building sites … 
hugo v0.126.1-3d40aba512931031921463dafc172c0d124437b8+extended linux/arm64 BuildDate=2024-05-15T10:42:34Z VendorInfo=gohugoio

ERROR Cannot find page: '/burnfix/' at "/home/vinser/Dev/Sites/hugo-hinode/site/content/en/apps/burnfix.md:24:61"
ERROR Cannot find page: 'https://vinser.github.io/flibgolite' at "/home/vinser/Dev/Sites/hugo-hinode/site/content/en/apps/flibgolite.md:28:61"
ERROR Cannot find page: 'https://vinser.github.io/pixyne' at "/home/vinser/Dev/Sites/hugo-hinode/site/content/en/apps/pixyne.md:27:61"
ERROR Cannot find page: '/burnfix/' at "/home/vinser/Dev/Sites/hugo-hinode/site/content/en/apps/burnfix.md:21:59"
ERROR Cannot find page: 'https://vinser.github.io/pixyne' at "/home/vinser/Dev/Sites/hugo-hinode/site/content/en/apps/pixyne.md:25:4"
ERROR Cannot find page: 'https://vinser.github.io/flibgolite' at "/home/vinser/Dev/Sites/hugo-hinode/site/content/en/apps/flibgolite.md:24:90"
Total in 1774 ms
Error: error building site: logged 6 error(s)
```
I can't follow your advice to use relative links due to the fact that these links point and should point to external sites and in my case they are sites of separate projects :'( The site I'm testing now I shall use as my person site which has foreign links on three project sites

**Reply by markdumay** [reply, marked as answer]
Ah ok, got it! I thought you were referencing a local page generated by Hugo. The solution I can think of is to add a new argument `external=true` to the link shortcode. When set, Hinode would treat the link as an external link and will not try to load the Hugo page.

**Reply by vinser** [reply]
I analyzed the link shortcode and also came to the understanding that it would be nice to add a parameter. This will allow the developer to have more control over the use of links.

**Reply by vinser** [reply]
I've added new argument `external` into 3 files in my local repo:
1. data/structures/link.yml
2. layouts/shortcodes/link.html
3. layouts/partials/assets/link.html

and now it runs as I expected.

Thanks for your suggestion! Kind regards

PS To be honest, I didn’t understand why all this curly cutting of the baseURL from a link was done. IMO `http` at link start was enough to separate local and external links. But now for backward compatibility you cannot do without an additional parameter.

**Reply by markdumay** [reply]
> PS To be honest, I didn’t understand why all this curly cutting of the baseURL from a link was done. IMO http at link start was enough to separate local and external links. But now for backward compatibility you cannot do without an additional parameter.

If I remember correctly, this was introduced to enable branch deployments (see the [versioning guide)](https://gethinode.com/guides/versioning/). But the code may have become a bit too complicated... :-)

**Reply by markdumay** [reply]
> I've added new argument external into 3 files in my local repo

Would you be willing to submit a PR with the proposed changes?

**Comment by vinser** [comment]
> > I've added new argument external into 3 files in my local repo
> 
> Would you be willing to submit a PR with the proposed changes?

I'll do my best

**Reply by vinser** [reply]
I've made PR to hinode theme. 
I did not make PR with example to docs site.  
Only brief comment on the new param in theme data/structures/link.yml will be shown in documentation.

**Reply by markdumay** [reply]
Thanks, it has been merged! I'll release a new version of Hinode next.
